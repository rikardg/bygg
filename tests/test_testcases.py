from dataclasses import dataclass
from pathlib import Path
import subprocess

import pytest


@dataclass
class TestcaseParameters:
    __test__ = False
    name: str
    actions: list[str]


testcases_dir = Path("testcases")

testcases: list[TestcaseParameters] = [
    TestcaseParameters("action_set", ["testfiles_static"]),
]


filetree_ignore = (
    "__pycache__",
    ".bygg",
)


def filetree(path: str):
    return sorted(
        [
            str(p.relative_to(path))
            for p in Path(path).rglob("**/*")
            if not any(sub in str(p) for sub in filetree_ignore)
        ]
    )


@pytest.mark.parametrize("testcase", testcases, ids=lambda x: x.name)
def test_list(snapshot, clean_bygg_tree, testcase):
    example_path = clean_bygg_tree / testcases_dir / testcase.name

    # assert filetree_before == snapshot

    process = subprocess.run(
        ["bygg", "--list"],
        cwd=clean_bygg_tree / testcases_dir / testcase.name,
        capture_output=True,
        encoding="utf-8",
    )
    assert process.returncode == 0
    assert filetree(example_path) == snapshot

    for action in testcase.actions:
        process = subprocess.run(
            ["bygg", action],
            cwd=clean_bygg_tree / testcases_dir / testcase.name,
            capture_output=True,
            encoding="utf-8",
        )
        assert process.returncode == 0
        assert filetree(example_path) == snapshot

        process = subprocess.run(
            ["bygg", action, "--clean"],
            cwd=clean_bygg_tree / testcases_dir / testcase.name,
            capture_output=True,
            encoding="utf-8",
        )
        assert process.returncode == 0
        assert filetree(example_path) == snapshot
