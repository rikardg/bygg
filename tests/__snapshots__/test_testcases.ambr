# serializer version: 1
# name: test_list[action_set]
  list([
    'Byggfile.py',
    'files',
    'files/a.txt',
    'files/b.txt',
    'files/c.txt',
  ])
# ---
# name: test_list[action_set].1
  list([
    'Byggfile.py',
    'files',
    'files/a.txt',
    'files/b.txt',
    'files/c.txt',
    'files/out',
    'files/out/a.txt',
    'files/out/b.txt',
    'files/out/c.txt',
  ])
# ---
# name: test_list[action_set].2
  list([
    'Byggfile.py',
    'files',
    'files/a.txt',
    'files/b.txt',
    'files/c.txt',
    'files/out',
  ])
# ---
